test/framework/src/main/java/org/elasticsearch/common/logging/JsonLogs.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.elasticsearch.common.xcontent.DeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+
+import java.io.BufferedReader;
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Iterator;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class JsonLogs implements Iterable<JsonLogLine>, Closeable {
```

Comment:
Could use javadoc I think.

Reply:
I think `Iterable`s have to support calling `iterator` multiple times to return independent `Iterator`. I don't see it in the Javadocs for `Iterable` but I'd be confused by an `Iterable` that works like this one. It might be more kind to have it implement `Stream` because that has `close` and "run once" semantics.

Reply:
agree -that needs a javadoc & and a cleanup.
The main intention is to use that class for readable asertions for json logger tests
That is why it has to be Iterable. You could think of this class as stream or Scanner, but neither of those has support in hamcrest at the version we use.

It can't only be stream, as we have some tests at the moment where the logs is traversed, some decision is made and then continued to be traversed from that point. We can't do that in stream (run once)

Refactored this to use a Reader supplier. Every time a new stream or iterator is created, a new Reader is created from the supplier.